File: src/main/java/model/IPEntry.java
package model;

import java.text.ParseException;

public class IPEntry {

    private final IPRange range;

    private final String countryCode;
    private final String country;
    private final String state;
    private final String city;

    private final Geolocation geolocation;

    public static IPEntry parseLine(String line) throws ParseException {
        var entries = line.split(",");

        if (entries.length < 8)
            throw new ParseException("Invalid data input: "+line, 0);

        long ipStart = Long.parseLong(entries[0]);
        long ipEnd = Long.parseLong(entries[1]);
        var ipRange = new IPRange(ipStart, ipEnd);

        String countryCode = entries[2];
        String country = entries[3];
        String state = entries[4];
        String city = entries[entries.length - 3];

        float lattit = Float.parseFloat(entries[entries.length - 2]);
        float longit = Float.parseFloat(entries[entries.length - 1]);
        var geoloc = new Geolocation(lattit, longit);

        return new IPEntry(
                ipRange,
                countryCode, country, state, city,
                geoloc);
    }

    public IPEntry(
            IPRange range,
            String countryCode, String country, String state, String city,
            Geolocation geolocation) {
        this.range = range;
        this.countryCode = countryCode;
        this.country = country;
        this.state = state;
        this.city = city;
        this.geolocation = geolocation;
    }

    public IPRange getRange() {
        return range;
    }

    public String getCountryCode() {
        return countryCode;
    }

    public String getCountry() {
        return country;
    }

    public String getState() {
        return state;
    }

    public String getCity() {
        return city;
    }

    public Geolocation getGeolocation() {
        return geolocation;
    }
}

File: src/model/ipentry.go
package model

import (
	"fmt"
	"strconv"
	"strings"
)


type IPEntry struct {
	Range IPRange
	//CountryCode, Country, State, City string
	Country string
	Location Geolocation
}

func ParseLine(line string) (*IPEntry, error) {
	entries := strings.Split(line, ",")
	if (len(entries) < 8) {
		return nil, fmt.Errorf("invalid data input: %s", line)
	}

	var startIP, endIP uint64
	startIP, _ = strconv.ParseUint(entries[0], 10, 32)
	endIP, _ = strconv.ParseUint(entries[1], 10, 32)

	var lat, long float64
	lat, _ = strconv.ParseFloat(entries[len(entries) - 2], 32)
	long, _ = strconv.ParseFloat(entries[len(entries) - 1], 32)

	ipentry := IPEntry{
		Range: IPRange{
			StartIP: uint32(startIP), 
			EndIP: uint32(endIP),
		},
		//CountryCode: entries[2],
		Country: entries[3],
		//State: entries[4],
		//City: entries[len(entries)-3],
		Location: Geolocation{
			Latitude: float32(lat),
			Longitude: float32(long),
		},
	}

	return &ipentry, nil
}
